Doc/library/functions.rst

```diff
@@ -1845,6 +1845,21 @@ are always available.  They are listed here in alphabetical order.
    :func:`super`, see `guide to using super()
    <https://rhettinger.wordpress.com/2011/05/26/super-considered-super/>`_.
 
+   .. note::
+      By default, if no second argument is provided, :func:`super` uses the first argument
+      of the immediately enclosing function definition, which is ``self`` under normal circumstances.
+      This may lead to unexpected behaviour if used inside nested functions or generator expressions
+      (the latter creates implicit nested functions). The following example will result in an exception::
+
+         class B(C):
+            def method(self, arg):
+               list(super().method(arg) for _ in range(1))   # Raises TypeError.
+
+      In such cases, you should provide arguments explicitly::
+
+         class B(C):
+            def method(self, arg):
+               list(super(B, self).method(arg) for _ in range(1))
```

Comment:
I think the clarification of what zero-argument super is and how it works should a) be shorter and more streamlined than this, and b) belongs higher up in the flow. One reason the current doc is awkward is that the "typical use" example shows zero-arg super even though the prior text never mentioned or explained it at all; the current PR doesn't fix that awkwardness.

I also don't think the generator expression case is important enough to deserve so much vertical space and its own entire pair of extra code examples. A simple mention is plenty.

Here is how I would propose to modify the docs: https://gist.github.com/carljm/098a8ed9ab387f244c1e15a16bc193db

Reply:
First version was right after first code example but @AlexWaygood suggested this belonged to the bottom. Also, why is it problematic to have a code example? It is always a lot better to have a visual example that goes along with textual explanation IMHO.

Reply:
@WolframAlph I agree that in general examples make a concept easier to understand. But there are many different concepts involved in `super()`, and a large array of possible code examples we could show to illustrate different usages. We have a limited "budget" for users' attention in the docs, and we have to prioritize. It doesn't make sense to me that two out of three total code examples in the `super()` documentation would be dedicated to illustrating the very specific edge case of usage in generator expressions. It's better to focus their attention on the single already-existing example demonstrating the most common and best usage, rather than dedicating code examples to showing what will not work.

Reply:
👍 , updated MR